source/common/io/io_uring_worker.h

```diff
@@ -0,0 +1,509 @@
+#pragma once
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/logger.h"
+#include "source/common/io/io_uring.h"
+#include "source/common/io/io_uring_impl.h"
+
+#include "io_uring.h"
+
+namespace Envoy {
+namespace Io {
+
+class IoUringAcceptSocket : public IoUringSocket, protected Logger::Loggable<Logger::Id::io> {
+public:
+  IoUringAcceptSocket(os_fd_t fd, IoUringHandler& io_uring_handler, IoUringWorker& parent)
+      : fd_(fd), io_uring_handler_(io_uring_handler), parent_(parent) {}
+
+  // IoUringSocket
+  os_fd_t fd() const override { return fd_; }
+  void start() override;
+  void close() override;
+  void enable() override;
+  void disable() override;
+
+  void onAccept(int32_t result) override;
+  void onClose(int32_t result) override;
+  void onCancel(int32_t result) override;
+
+private:
+  os_fd_t fd_;
+
+  IoUringHandler& io_uring_handler_;
+  IoUringWorker& parent_;
+
+  sockaddr_storage remote_addr_;
+  socklen_t remote_addr_len_{sizeof(remote_addr_)};
+  os_fd_t connection_fd_{INVALID_SOCKET};
+
+  Request* accept_req_{nullptr};
+  Request* cancel_req_{nullptr};
+  Request* close_req_{nullptr};
+
+  bool is_disabled_{false};
+  bool is_pending_accept_{false};
+  bool is_closing_{false};
+};
+
+class IoUringBaseSocket : public IoUringSocket, protected Logger::Loggable<Logger::Id::io> {
+public:
+  IoUringBaseSocket(os_fd_t fd, IoUringHandler& io_uring_handler, IoUringWorker& parent,
+                    uint32_t read_buffer_size)
+      : fd_(fd), io_uring_handler_(io_uring_handler), parent_(parent),
+        read_buffer_size_(read_buffer_size), iov_(new struct iovec[1]) {}
+
+  virtual ~IoUringBaseSocket() {
+    if (SOCKET_VALID(fd_)) {
+      ::close(fd_);
+    }
+  }
+
+  // IoUringSocket
+  os_fd_t fd() const override { return fd_; }
+
+  void start() override {
+    ENVOY_LOG(debug, "start, fd = {}", fd_);
+    if (read_req_ == nullptr) {
```

Comment:
I think in most the cases, after we create the socket, then we will start to read/write the data. If we submit the read request first after creating the socket, then when we want to read later, the data may already be ready. I thought this will be more efficient.

Otherwise, we can submit the read req lazily. We can submit the read request when the `Iohandle::read` is invoked, but that means we need to return -EAGAIN on the first `Iohandle::read` invoke, and wait for iouring to get read finished. After the read request is done, we trigger the read event again, then wait for the up layer invoke `ioHandle::read` again to fetch the data.